### ISSUE TITLE
[Bug]: Babel generate incorrect source-map sources includes ??ref-- 

### ISSUE THREAD

**Issue by CYJB** [babel/babel#14676]
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

babel-loader (webpack)

### Input code

Any webpack project use a custom-loader with options before ts-loader, set devtool to source-map.

webpack.config.js is like below:

```js
const path = require('path');

module.exports = {
  entry: './src/index.ts',
  output: {
    path: path.resolve(__dirname, 'dist'),
    filename: 'bundle.js',
  },
  mode: 'production',
  devtool: "source-map",
  module: {
    rules: [{
      test: /\.ts$/,
      use: [
        { loader: 'babel-loader' },
        { loader: 'ts-loader', options: { compilerOptions: { target: 'esnext' } } },
        { loader: 'custom-loader', options: { anyKey: 'anyValue' } },
      ],
    }],
  },
};
```

### Current and expected behavior

Webpack will generate ident like `??ref--16-2` to custom-loader, then ts-loader will generate source-map sources like `/my-project/node_modules/custom-loader@0.0.1/index.js??ref--16-2!/my-project/src/index.ts`, babel generate source-map like: 

```json
{
	"version": 3,
	"file": "bundle.js",
	"mappings": "AAAAA,QAAQC,IAAI",
	"sources": [
		"webpack://my-project/./custom-loader/index.js"
	],
	"sourcesContent": [
		"console.log('test');\n"
	],
	"names": [
		"console",
		"log"
	],
	"sourceRoot": ""
}
```

sources is path to custom-loader but not src/index.ts

### Environment

- Babel version: 7.18.5
- Node: v16.14.0
- npm version: 8.3.1
- OS: macOS 12.1

**Comment by nicolo-ribaudo** [comment]
Hey, could you share a repository that we can use to debug the issue? :pray: It would help a lot with fixing this.

**Comment by CYJB** [comment]
> Hey, could you share a repository that we can use to debug the issue? 🙏 It would help a lot with fixing this.

You can see https://github.com/CYJB/Babel-14676-sourcemap-test

**Comment by nicolo-ribaudo** [comment]
@jridgewell This is a regression in `@jridgewell/trace-mapping@0.3.13`. It can be reproduced with this code:
<details>

```js
import remapping from "@ampproject/remapping";

// https://github.com/babel/babel/blob/b7631ffc6568818416ebfdd68403caa77bf63ac6/packages/babel-core/src/transformation/file/merge-map.ts#L4
function mergeSourceMap(inputMap, map, sourceFileName) {
  let found = false;
  return remapping(map, (s, ctx) => {
    if (s === sourceFileName && !found) {
      found = true;
      ctx.source = "";
      return inputMap;
    }

    return null;
  });
}

const inputMap = {
  version: 3,
  file: "/src/index.ts",
  sourceRoot: null,
  sources: [
    "/custom-loader/index.js??ruleSet[1].rules[0].use[2]!/src/index.ts",
  ],
  names: [],
  mappings: "AAAA,OAAO,CAAC,GAAG,CAAC,MAAM,CAAC,CAAC",
  sourcesContent: ["console.log('test');\n"],
};

const map = {
  version: 3,
  names: ["console", "log"],
  sourceRoot: null,
  sources: [
    "/src/index.ts",
  ],
  sourcesContent: ["console.log('test');\n"],
  mappings: [
    [
      [0, 0, 0, 0, 0],
      [7, 0, 0, 7],
      [8, 0, 0, 8, 1],
      [11, 0, 0, 0],
      [12, 0, 0, 12],
      [18, 0, 0, 0],
    ],
  ],
};

const sourceFileName = "/src/index.ts";

console.log(mergeSourceMap(inputMap, map, sourceFileName));
```

You can see the different behavior when using `@jridgewell/trace-mapping@0.3.12` and `@jridgewell/trace-mapping@0.3.13`.

</details>

@CYJB As a workaround, make sure that the version of `@jridgewell/trace-mapping` in your `node_modules` is `0.3.12` (it's a deep transitive dependency of Babel). Adding it explicitly to your `devDependencies` and re-running `npm install` will probably make it work, otherwise you'll need https://www.npmjs.com/package/npm-force-resolutions.

**Comment by jridgewell** [comment]
> This is a regression in `@jridgewell/trace-mapping@0.3.13`.

I agree this is a regression (in that the behavior changed), but it's a bugfix. `webpack://foo/./bar` is equivalent to `webpack://foo/bar` when using URLs, which is how sourcemap sources are treated. Further, if this sourcemap were to have a `sourceRoot`, this is the output that would occur because the resolution step normalizes the input.

I think what's actually the matter here is that I don't support `?query` params in the resolution, they're silently dropped. I can add that, which will result in sources like `webpack://foo/custom-loader@0.0.1/index.js??ref--16-2!/my-project/src/index.ts`. But, I have no idea what `webpack://` protocol is supposed to do, so I don't know how this is supposed to be useable by devtools.

**Comment by jridgewell** [comment]
Manually running https://github.com/babel/babel/issues/14676#issuecomment-1166932925 with an edited `@jridgewell/resolve-uri` that propagates the query param seems to fix the sourcemap output by webpack:

```json
{
  "version": 3,
  "file": "bundle.js",
  "mappings": "AAAAA,QAAQC,IAAI",
  "sources": [
    "webpack://sourcemap-test/./src/index.ts"
  ],
  "sourcesContent": [
    "console.log('test');\n"
  ],
  "names": [
    "console",
    "log"
  ],
  "sourceRoot": ""
}
```

That `webpack://...` isn't normalized because it's coming out of the query query param (which isn't normalized by URLs). Whatever sourcemap logic webpack must have is extracting and replacing it. The sourcemap being output by Babel with this change is:

```json
{
  "version": 3,
  "mappings": "AAAAA,OAAO,CAACC,GAAR,CAAY,MAAZ",
  "names": [ "console", "log" ],
  "sources": [
    "/Users/jridgewell/Downloads/Babel-14676-sourcemap-test-main/custom-loader/index.js??ruleSet[1].rules[0].use[2]!/Users/jridgewell/Downloads/Babel-14676-sourcemap-test-main/src/index.ts"
  ],
  "sourcesContent": [ "console.log('test');\n" ]
}
```

**Comment by jridgewell** [comment]
This should be resolved by `@jridgewell/resolve-uri` [v3.1.0](https://github.com/jridgewell/resolve-uri/releases/tag/v3.1.0). You'll need to upgrade the transitive dependency by running `npm upgrade`

**Comment by nicolo-ribaudo** [comment]
Thanks @jridgewell!
